Sources/Subprocess/AsyncBufferSequence.swift

```diff
@@ -74,6 +80,259 @@ public struct AsyncBufferSequence: AsyncSequence, Sendable {
+            private var startIndex: Int
+            private let bufferingPolicy: BufferingPolicy
+
+            internal init(
+                underlyingIterator: AsyncBufferSequence.AsyncIterator,
+                bufferingPolicy: BufferingPolicy
+            ) {
+                self.source = underlyingIterator
+                self.buffer = []
+                self.eofReached = false
+                self.startIndex = 0
+                self.bufferingPolicy = bufferingPolicy
+            }
+
+            public mutating func next() async throws -> String? {
+
+                func loadBuffer() async throws -> [Encoding.CodeUnit]? {
+                    guard !self.eofReached else {
+                        return nil
+                    }
+
+                    guard let buffer = try await self.source.next() else {
+                        self.eofReached = true
+                        return nil
+                    }
+                    #if os(Windows)
+                    // Cast data to CodeUnit type
+                    let result = buffer.withUnsafeBytes { ptr in
+                        return Array(
+                            UnsafeBufferPointer<Encoding.CodeUnit>(
+                                start: ptr.bindMemory(to: Encoding.CodeUnit.self).baseAddress!,
+                                count: ptr.count / MemoryLayout<Encoding.CodeUnit>.size
+                            )
+                        )
+                    }
+                    #else
+                    // Unfortunately here we _have to_ copy the bytes out because
+                    // DispatchIO (rightfully) reuses buffer, which means `buffer.data`
+                    // has the same address on all iterations, therefore we can't directly
+                    // create the result array from buffer.data
+
+                    // Calculate how many CodePoint elements we have
+                    let elementCount = buffer.data.count / MemoryLayout<Encoding.CodeUnit>.stride
+
+                    // Create array by copying from the buffer reinterpreted as CodePoint
+                    let result: Array<Encoding.CodeUnit> = buffer.data.withUnsafeBytes { ptr -> Array<Encoding.CodeUnit> in
+                        return Array(
+                            UnsafeBufferPointer(start: ptr.baseAddress?.assumingMemoryBound(to: Encoding.CodeUnit.self), count: elementCount)
+                        )
+                    }
+
+                    #endif
+                    return result.isEmpty ? nil : result
+                }
+
+                func yield(at endIndex: Int) -> String? {
+                    defer {
+                        self.buffer.removeFirst(endIndex)
+                        self.startIndex = 0
+                    }
+                    if self.buffer.isEmpty {
+                        return nil
+                    }
+                    return String(decoding: self.buffer[0 ..< endIndex], as: Encoding.self)
```

Comment:
I'm still a little unclear on why we need to copy this into an Array, and then copy it into a String. Why can't String read from the buffer data directly?

Reply:
We have to keep an `[UInt8]` buffer in order to scan through and find new line characters. We can't use `String` directly as the buffer type because every time when we receive a chunk of data, we can't directly convert it to String since the chunk might end in between byte encodings.